c10/core/impl/LocalDispatchKeySet.h

```diff
@@ -95,28 +85,85 @@ class C10_API IncludeDispatchKeyGuard {
  private:
   // A little micro-optimization to save us from tls_get_addr call
   // on destruction
-  PODLocalDispatchKeySet* tls_;
+  LocalDispatchKeySetWrapper tls_wrapper_;
   DispatchKeySet include_;
 };
 
-class C10_API ExcludeDispatchKeyGuard {
+template <uint64_t exclude, bool has_overlap>
+class ExcludeDispatchKeyGuard {
  public:
-  ExcludeDispatchKeyGuard(DispatchKeySet);
-  ExcludeDispatchKeyGuard(DispatchKey k)
-      : ExcludeDispatchKeyGuard(DispatchKeySet(k)) {}
+  // If our exclude set does not overlap with c10::default_excluded_set, we can
+  // skip some bookkeeping. (And we know at compile time if this is the case.)
+  // Key exclusion tends to be on the hot path, so it's worth it to bypass the
+  // unnecessary XORs if possible. We force `has_overlap` to be explicitly
+  // declared so that changes to the default excluded set do not silently knock
+  // us off the hot path.
```

Comment:
I'm not convinced by the explicit `has_overlap` argument here. Suppose that someone makes a change to default excluded set, and it is valid to remove things off the hot path. Is the point here to just to force someone to notice, in code review, that we are no longer on the hot path? It seems like you could do that a lot more straightforwardly by just having a sequence of static asserts that such and such dispatch key is in (not) in the default excluded set. This would still have the desired path of flagging such a change in review... while not simultaneously forcing the person to also change all of the call sites to update the (redundant) has overlap information.

Reply:
+1, I think this encumbers the API a bit too much, in addition to making changes potentially higher friction than they need to be. Taylor do you see a problem with going with Ed's suggestion of just static asserting the keys we care about directly instead?

Reply:
Yeah, I think that is very reasonable. I think this also lets us get rid of the specialize macro altogether.